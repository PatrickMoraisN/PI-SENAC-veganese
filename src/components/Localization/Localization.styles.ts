import styled from 'styled-components'

type ButtonProps = {
  error: boolean
}

export const Button = styled.button<ButtonProps>`
  padding: 8px;
  background: #e6e5e5;
  border-radius: 6px;
  height: 32px;
  display: flex;
  justify-content: center;
  align-items: center;
  border: none;
  color: #574f4d;
  cursor: pointer;
  transition: 0.3s;
  border: ${(props) => props.error && '3px solid red'};

  &:hover {
    filter: brightness(0.9);
  }
`
