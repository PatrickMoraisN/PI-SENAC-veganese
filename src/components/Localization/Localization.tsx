import { calcCrow } from '@/api/shipping'
import { sede } from '@/enums/sede'
import React, { useState } from 'react'
import * as S from './Localization.styles'

type LocalizationErrorProps = string | null

type CoordsProps = {
  latitude: number
  longitude: number
}

type PositionCoordsProps = {
  coords: CoordsProps
}

type LocalizationProps = {
  setShipping: (value: number) => void
  setShippingCalculated: (value: boolean) => void
  error: string | null
}

export function Localization({
  setShipping,
  setShippingCalculated,
  error,
}: LocalizationProps) {
  const [coords, setCoords] = useState<CoordsProps | null>(null)
  const [localizationError, setLocalizationError] =
    useState<LocalizationErrorProps>(null)

  async function getLocation() {
    try {
      if (navigator.geolocation) {
        navigator.geolocation.getCurrentPosition(showPosition)
      } else {
        setLocalizationError('Geolocation is not supported by this browser.')
      }
    } catch (err) {
      setLocalizationError('Error.')
    }
  }

  function showPosition(position: PositionCoordsProps) {
    setCoords({
      latitude: position.coords.latitude,
      longitude: position.coords.longitude,
    })
  }

  function calculateShip() {
    const { latitude, longitude } = sede
    const { latitude: l2, longitude: lo2 } = coords
    const distance = calcCrow(latitude, longitude, Number(l2), Number(lo2))
    return distance
  }

  function getShipping() {
    const shipping = calculateShip() * 2
    setShipping(shipping)
    setShippingCalculated(true)
    return shipping.toFixed(2)
  }

  function shippingError() {
    return error === 'O Frete não está calculado'
  }

  return (
    <>
      {!coords && !localizationError && (
        <S.Button onClick={getLocation} error={shippingError()}>
          Calcular Frete
        </S.Button>
      )}
      {localizationError && <p>Erro ao calcular!</p>}
      {coords && !localizationError && (
        <>
          <p>Entrega</p>
          <p>R$ {getShipping()}</p>
        </>
      )}
    </>
  )
}
