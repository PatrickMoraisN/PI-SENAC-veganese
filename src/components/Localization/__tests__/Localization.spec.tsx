import userEvent from '@testing-library/user-event'
import { render, screen, waitFor } from '@testing-library/react'
import { Localization } from '../Localization'

const setShipping = jest.fn()
const setShippingCalculated = jest.fn()

const renderLocalization = () =>
  render(
    <Localization
      setShipping={setShipping}
      error={null}
      setShippingCalculated={setShippingCalculated}
    />,
  )

describe('Localization', () => {
  afterEach(() => {
    jest.clearAllMocks()
  })

  it('should render correctly', () => {
    const { container } = renderLocalization()

    expect(container).toMatchSnapshot()
  })

  it('should render shipping correctly', async () => {
    const mockGeolocation = {
      getCurrentPosition: jest.fn().mockImplementation((success) =>
        Promise.resolve(
          success({
            coords: {
              latitude: 10,
              longitude: 10,
            },
          }),
        ),
      ),
    }
    // @ts-ignore
    navigator.geolocation = mockGeolocation
    renderLocalization()

    await userEvent.click(screen.getByText('Calcular Frete'))

    expect(() => screen.getByText('Calcular Frete')).toThrow()
    expect(screen.getByText('Entrega')).toBeInTheDocument()
  })

  it('should render shipping correctly', async () => {
    const mockGeolocation = {
      getCurrentPosition: jest
        .fn()
        .mockImplementationOnce((_successCallback, errorCallback) => {
          errorCallback(new Error('w'))
        }),
    }

    // @ts-ignore
    navigator.geolocation = mockGeolocation
    renderLocalization()

    await userEvent.click(screen.getByText('Calcular Frete'))

    expect(() => screen.getByText('Calcular Frete')).toThrow()
    await waitFor(() =>
      expect(screen.getByText('Erro ao calcular!')).toBeInTheDocument(),
    )
  })
})
