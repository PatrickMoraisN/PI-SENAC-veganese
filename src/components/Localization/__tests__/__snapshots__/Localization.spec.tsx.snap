// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Localization should render correctly 1`] = `
<div>
  <button
    class="sc-bcXHqe kyvJqT"
  >
    Calcular Frete
  </button>
</div>
`;
